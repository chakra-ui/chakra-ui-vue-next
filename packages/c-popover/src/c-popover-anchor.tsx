import { defineComponent } from "vue"
import { usePopoverContext } from "./popover.context"
import { HTMLChakraProps, chakra } from "@chakra-ui/vue-system"

export interface CPopoverAnchorProps extends HTMLChakraProps<"div"> {}
export const CPopoverAnchor = defineComponent({
  name: "CPopoverAnchor",
  setup(_, { slots, attrs }) {
    const api = usePopoverContext()

    return () => (
      <chakra.div {...api.value.anchorProps} {...attrs}>
        {slots.default?.()}
      </chakra.div>
    )
  },
})
