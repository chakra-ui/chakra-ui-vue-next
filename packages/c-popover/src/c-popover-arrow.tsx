import { computed, defineComponent, watchEffect } from "vue"
import { usePopoverContext, useStyles } from "./popover.context"
import { HTMLChakraProps, chakra } from "@chakra-ui/vue-system"

export interface CPopoverArrow extends HTMLChakraProps<"div"> {}
export const CPopoverArrow = defineComponent({
  name: "CPopoverArrow",
  inheritAttrs: false,
  setup(_, { slots, attrs }) {
    const api = usePopoverContext()

    const styles = useStyles()

    const arrowBg = computed(
      () => attrs.bg ?? attrs.bgColor ?? attrs.backgroundColor
    )
    const arrowShadow = computed(() => attrs.shadow ?? attrs.boxShadow)

    return () => (
      <chakra.div
        __label="popover__arrow-positioner"
        {...api.value.arrowProps}
        __css={{
          "--arrow-size": "10px",
          "--popper-arrow-default-shadow":
            "-1px -1px 1px 0 var(--popper-arrow-shadow-color)",
          "--arrow-shadow-color": `${
            arrowShadow.value
              ? `shadows.${arrowShadow.value} ${arrowShadow.value}`
              : ""
          } var(--popper-arrow-shadow, var(--popper-arrow-default-shadow))`,
        }}
      >
        <chakra.div
          __css={{
            ...styles.value.arrow,
            "--arrow-size": "10px",
            "--arrow-background": arrowBg.value
              ? `colors.${arrowBg.value}, var(--popper-arrow-bg), ${arrowBg.value}`
              : "var(--popper-arrow-bg)",
            boxShadow:
              "var(--popper-arrow-shadow, var(--popper-arrow-default-shadow))",
          }}
          __label="popover__arrow"
          {...api.value.arrowTipProps}
          {...attrs}
        />
        {slots.default?.()}
      </chakra.div>
    )
  },
})
