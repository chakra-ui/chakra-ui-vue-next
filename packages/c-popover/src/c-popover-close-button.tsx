import { computed, defineComponent } from "vue"
import { usePopoverContext, useStyles } from "./popover.context"
import { CCloseButton, type CloseButtonProps } from "@chakra-ui/c-close-button"
import { match } from "@chakra-ui/vue-utils"

export interface CPopoverCloseButtonProps extends CloseButtonProps {}
export const CPopoverCloseButton = defineComponent({
  name: "CPopoverCloseButton",
  setup(_, { attrs }) {
    const api = usePopoverContext()
    const styles = useStyles()

    return () => (
      <CCloseButton
        size="sm"
        __label="popover__close-button"
        __css={styles.value.closeButton}
        {...api.value.triggerProps}
        {...attrs}
      />
    )
  },
})
