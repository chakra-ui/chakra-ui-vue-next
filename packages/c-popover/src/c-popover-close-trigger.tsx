import { defineComponent } from "vue"
import { usePopoverContext } from "./popover.context"
import { withSingleton } from "@chakra-ui/vue-utils"

export const CPopoverCloseTrigger = defineComponent({
  name: "CPopoverCloseTrigger",
  setup(_, { slots, attrs }) {
    const api = usePopoverContext()

    return () =>
      withSingleton(slots, "CPopoverCloseTrigger", {
        ...api.value.closeTriggerProps,
        ...attrs,
      })
  },
})
