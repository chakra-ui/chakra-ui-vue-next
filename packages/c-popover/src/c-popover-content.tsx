import { Transition, computed, defineComponent, watch, watchEffect } from "vue"
import { usePopoverContext, useStyles } from "./popover.context"
import {
  type HTMLChakraProps,
  type SystemStyleObject,
  chakra,
} from "@chakra-ui/vue-system"
import { CPopoverPositioner } from "./c-popover-positioner"
import { TransitionDefaults } from "@chakra-ui/c-motion"
import { MotionDirective, useMotions } from "@vueuse/motion"
import { withDirectives } from "vue"
import { match } from "@chakra-ui/vue-utils"

const variants = {
  backdrop: {
    initial: {
      opacity: 0,
      scale: 1.1,
    },
    animate: { opacity: 0.5, scale: 1 },
    exit: { opacity: 0, scale: 1.1 },
  },
  content: {
    initial: {
      opacity: 0,
      scale: 1.005,
      y: 10,
    },
    animate: { opacity: 1, scale: 1, y: 0 },
    exit: { opacity: 0, scale: 1.005, y: 10 },
  },
}

export interface CPopoverContentProps extends HTMLChakraProps<"div"> {}
export const CPopoverContent = defineComponent({
  name: "CPopoverContent",
  inheritAttrs: false,
  setup(_, { slots, attrs }) {
    const api = usePopoverContext()

    const styles = useStyles()
    const contentStyles = computed<SystemStyleObject>(() => ({
      position: "relative",
      display: "flex",
      flexDirection: "column",
      ...styles.value.content,
    }))

    const popoverContentProps = computed(() => {
      const { ...rest } = { ...attrs, ...api.value.contentProps }
      return {
        ...rest,
        ...match(api.value.trigger, {
          hover: {
            async onPointerenter(e: MouseEvent) {
              const motions = useMotions()
              const instance = motions[api.value.transitionId]
              instance.stopTransitions()
              api.value.open()
              requestAnimationFrame(() => {
                api.value.enterTransition(() => null)
              })
            },
            async onPointerleave(e: MouseEvent) {
              requestAnimationFrame(() => {
                api.value.leaveTransition(() => api.value.close())
              })
            },
          },
          click: {},
        }),
      }
    })

    return () => (
      <CPopoverPositioner>
        {withDirectives(
          <chakra.div
            {...popoverContentProps.value}
            __css={contentStyles.value}
            __label="popover__content"
          >
            {slots.default?.()}
          </chakra.div>,
          [
            [
              MotionDirective({
                initial: { scale: 0.95, opacity: 0 },
                enter: {
                  scale: 1,
                  transition: TransitionDefaults.enter,
                  opacity: 1,
                },
                leave: {
                  scale: 0.95,
                  transition: TransitionDefaults.leave,
                  opacity: 0,
                },
              }),
              api.value.transitionId,
            ],
          ]
        )}
      </CPopoverPositioner>
    )
  },
})
