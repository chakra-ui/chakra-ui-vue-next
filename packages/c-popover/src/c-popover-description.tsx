import { defineComponent } from "vue"
import { usePopoverContext } from "./popover.context"
import { HTMLChakraProps, chakra } from "@chakra-ui/vue-system"

export interface CPopoverDescriptionProps extends HTMLChakraProps<"div"> {}
export const CPopoverDescription = defineComponent({
  name: "CPopoverDescription",
  setup(_, { slots, attrs }) {
    const api = usePopoverContext()

    return () => (
      <chakra.div {...attrs} {...api.value.descriptionProps}>
        {slots.default?.()}
      </chakra.div>
    )
  },
})
