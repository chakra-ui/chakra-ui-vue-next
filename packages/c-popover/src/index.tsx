export { CPopover } from "./c-popover"
export { CPopoverContent, type CPopoverContentProps } from "./c-popover-content"
export {
  CPopoverPositioner,
  type CPopoverPositionerProps,
} from "./c-popover-positioner"
export { CPopoverTrigger } from "./c-popover-trigger"
export { CPopoverCloseTrigger } from "./c-popover-close-trigger"
export { CPopoverTitle, type CPopoverTitleProps } from "./c-popover-title"
export { CPopoverArrow } from "./c-popover-arrow"
export {
  CPopoverDescription,
  type CPopoverDescriptionProps,
} from "./c-popover-description"
export { CPopoverHeader, type CPopoverHeaderProps } from "./c-popover-header"
export { CPopoverFooter, type CPopoverFooterProps } from "./c-popover-footer"
export { CPopoverBody, type CPopoverBodyProps } from "./c-popover-body"
export { CPopoverAnchor, type CPopoverAnchorProps } from "./c-popover-anchor"
export {
  CPopoverCloseButton,
  type CPopoverCloseButtonProps,
} from "./c-popover-close-button"
export { usePopover } from "./use-popover"

export type { CPopoverProps } from "./c-popover"
export type { UsePopoverProps, UsePopoverPropsContext } from "./use-popover"
