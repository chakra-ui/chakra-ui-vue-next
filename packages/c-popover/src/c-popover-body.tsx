import { defineComponent } from "vue"
import { usePopoverContext } from "./popover.context"
import { HTMLChakraProps, chakra } from "@chakra-ui/vue-system"
import { CPopoverDescription } from "./c-popover-description"

export interface CPopoverBodyProps extends HTMLChakraProps<"div"> {}
export const CPopoverBody = defineComponent({
  name: "CPopoverBody",
  setup(_, { slots, attrs }) {
    const api = usePopoverContext()

    return () => (
      <CPopoverDescription {...attrs}>{slots.default?.()}</CPopoverDescription>
    )
  },
})
