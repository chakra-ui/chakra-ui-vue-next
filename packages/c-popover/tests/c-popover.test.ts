import { render } from "@chakra-ui/vue-test-utils"
import { CPopover } from "../src"

it("should render properly", () => {
  const { asFragment } = render(CPopover)
  expect(asFragment()).toMatchSnapshot()
})
