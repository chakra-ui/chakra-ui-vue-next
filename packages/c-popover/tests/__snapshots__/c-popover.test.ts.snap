// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render properly 1`] = `
<DocumentFragment>
  <!---->
</DocumentFragment>
`;
